lib/puppet/provider/package/pip.rb

```diff
@@ -124,35 +124,19 @@ def latest
     self.class.validate_command(command)
 
     command_version = self.class.pip_version(command)
-    if Puppet::Util::Package.versioncmp(command_version, '1.5.4') == -1
-      latest_with_old_pip
+    if self.class.compare_pip_versions(command_version, '1.5.4') == -1
+      available_versions_with_old_pip.last
     else
-      latest_with_new_pip
+      available_versions_with_new_pip.last
     end
   end
 
-  # Less resource-intensive approach for pip version 1.5.4 and newer.
-  def latest_with_new_pip
-    available_versions_with_new_pip.last
-  end
-
-  # More resource-intensive approach for pip version 1.5.3 and older.
-  def latest_with_old_pip
-    command = resource_or_provider_command
-    self.class.validate_command(command)
-
-    Dir.mktmpdir("puppet_pip") do |dir|
-      command_and_options = [command, 'install', "#{@resource[:name]}", '-d', "#{dir}", '-v']
-      command_and_options << install_options if @resource[:install_options]
-      execpipe command_and_options do |process|
-        process.collect do |line|
-          # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
-          end
-        end
-        return nil
-      end
+  def self.compare_pip_versions(x, y)
+    begin
+      Puppet::Util::Package::Version::Pip.compare(x, y)
+    rescue PIP_VERSION::ValidationFailure => ex
+      Puppet.debug("Cannot compare #{x} and #{y}. #{ex.message} Falling through default comparison mechanism.")
+      Puppet::Util::Package.versioncmp(x, y)
```

Comment:
Is there any test covering fallback to `Puppet::Util::Package.versioncmp`?

Reply:
Added one.